### ISSUE TITLE
Warn if pyOpenSSL <0.10 is used

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4974]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4974|
|Type|enhancement|
|Created|2011-03-21 21:43:05Z|
|Branch|https://github.com/twisted/twisted/tree/deprecate-oldtls-4974|

pyOpenSSL introduced memory BIO APIs.  These let Twisted do SSL better - faster and with fewer bugs, and with more consistent cross-platform behavior.  [#4854](https://github.com/twisted/twisted/issues/4854) switched `IReactorSSL` implementations to prefer these APIs, but still fall back to the old socket BIO APIs.  Still, we should encourage people to move forward.  They will get better behavior and we will be able to drop some gross code.

So, if pyOpenSSL 0.10 or newer is not available on a system, we should indicate our preference for a newer version, probably with a `UserWarning` (unless anyone has a suggestion for something better).

<details><summary>Searchable metadata</summary>

```
trac-id__4974 4974
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__totally_automated_release_infrastructure totally automated release infrastructure
branch__branches_deprecate_oldtls_4974 branches/deprecate-oldtls-4974
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1300743785000000 1300743785000000
changetime__1336817271000000 1336817271000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

See [#5014](https://github.com/twisted/twisted/issues/5014) for what comes after this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [31537]) Merge protocol-ssl-4854-6 

Author: exarkun
Reviewer: itamar, glyph, thijs
Fixes: [#4854](https://github.com/twisted/twisted/issues/4854)
Refs: [#4974](https://github.com/twisted/twisted/issues/4974)
Refs: [#5014](https://github.com/twisted/twisted/issues/5014)
Refs: [#4455](https://github.com/twisted/twisted/issues/4455)

Add an implementation of `IReactorSSL` and `ITLSTransport` which uses the memory
BIO APIs present in pyOpenSSL 0.10 and newer.  This implementation will be preferred
by all reactors if the pyOpenSSL dependency is satisfied, otherwise the old
implementation will still be used.

This appears to have slightly better performance than the old implementation and
should avoid bugs like [#4455](https://github.com/twisted/twisted/issues/4455).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

This is not going to happen in 11.1.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [34341]) Branching to 'deprecate-oldtls-4974'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

 * [Debian Stable has 0.10](http://packages.debian.org/search?keywords=python+openssl&searchon=names&suite=stable&section=all)
 * [Ubuntu Lucid has 0.10](http://packages.ubuntu.com/search?keywords=python+openssl&searchon=names&suite=lucid&section=all)
 * [Fedora 15 has 0.12](http://pkgs.fedoraproject.org/gitweb/?p=pyOpenSSL.git;a=commit;h=742a86067e9b8fc693cf458f20649d6fb5193d56)

Determining the version of a package available in RHEL appears to [require a redhat network ID](http://unix.stackexchange.com/questions/6864/how-to-search-for-official-rhel-packages), which seems to cost money, and which I do not have.  So RHEL loses out if they're still on an old version of pyOpenSSL, as far as I'm concerned.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

Also, [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/deprecate-oldtls-4974).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst set owner to @exarkun|
|-|-|

Looks good, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [34360]) Merge deprecate-oldtls-4974

Author: exarkun
Reviewer: itamar
Fixes: [#4974](https://github.com/twisted/twisted/issues/4974)

Deprecate support for pyOpenSSL versions older than 0.10, which lack support
for in-memory operation.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `reopened`|
|-|-|

(In [34361]) Revert r34360 - test suite regression

Reopens: [#4974](https://github.com/twisted/twisted/issues/4974)

Numerous issues on various platforms.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [34363]) Skip the test if OpenSSL is missing

refs [#4974](https://github.com/twisted/twisted/issues/4974)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Handled the two other cases that caused test failures:

  1. OpenSSL is not present at all and
  1. an old version of OpenSSL is present causing `_oldtls` to have been loaded already

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/deprecate-oldtls-4974) look better now, though the FreeBSD slave is offline at this time.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_13"></a>|@itamarst set owner to @exarkun|
|-|-|

Please (re)merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set status to `closed`|
|-|-|

(In [34370]) Merge deprecate-oldtls-4974 (again)

Author: exarkun
Reviewer: itamar
Fixes: [#4974](https://github.com/twisted/twisted/issues/4974)

Deprecate support for pyOpenSSL versions older than 0.10, which lack support
for in-memory operation.

Re-merge fixes shortcomings of the first attempt, which led to test failures
when OpenSSL was not installed or when the installed version of OpenSSL
was older than 0.10.
